Docs feature branch: SQLite in DO (#16903)
cloudflare/cloudflare-docs: src/content/docs/durable-objects/api/storage-api.mdx

#### Supported options

* `setAlarm()` and `deleteAlarm()` support the same options as [`put()`](/durable-objects/api/storage-api/#put), but without `noCache`.

### sql.exec

:::note[SQLite in Durable Objects Beta]

SQL API methods accessed with `ctx.storage.sql` are only allowed on [Durable Object classes with SQL storage](/durable-objects/reference/durable-objects-migrations/#enable-sql-storage-on-create-durable-object-class-migration) and will return an error if called on Durable Objects with key-value storage.

:::

<code>ctx.storage.sql.exec(query: string, ...bindings: any[])</code> : SqlStorageCursor

#### Parameters
* `query`: string 
  * The SQL query string to be executed. `query` can contain `?` placeholders for parameter bindings.
* `bindings`: any[] Optional
  * Optional variable number of arguments that correspond to the `?` placeholders in `query`.

#### Returns 

> A cursor (`SqlStorageCursor`) to iterate over query row results as objects. Cursor is a JavaScript [Iterable](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_iterable_protocol), which supports iteration using `for (let row of cursor)`. Cursor is also a JavaScript [Iterator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_iterator_protocol), which supports iteration using `cursor.next()`.

Author: cc @justin-mp @kentonv for another pass of `sql.exec` api docs

> A cursor (`SqlStorageCursor`) to iterate over query row results as objects. Cursor is a JavaScript [Iterable](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_iterable_protocol), which supports iteration using `for (let row of cursor)`. Cursor is also a JavaScript [Iterator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_iterator_protocol), which supports iteration using `cursor.next()`.

Reviewer: ```suggestion
A cursor (`SqlStorageCursor`) to iterate over query row results as objects. `SqlStorageCursor` is a JavaScript [Iterable](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_iterable_protocol), which supports iteration using `for (let row of cursor)`. `SqlStorageCursor` is also a JavaScript [Iterator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_iterator_protocol), which supports iteration using `cursor.next()`.
```

> Returned cursor supports the following methods:

Reviewer: ```suggestion
`SqlStorageCursor` supports the following methods:
```

  
* `next()`

>   * Returns an object representing the next value of the cursor. Object has `done` and `value` properties. `done` is set to `false` when a next value is present, and `value` is set to the next row object in the query result. `done` is set to `true` when the entire cursor is consumed, and no `value` is set. 

Reviewer: ```suggestion
  * Returns an object representing the next value of the cursor. The returned object has  `done` and `value` properties adhering to the JavaScript [Iterator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_iterator_protocol). `done` is set to `false` when a next value is present, and `value` is set to the next row object in the query result. `done` is set to `true` when the entire cursor is consumed, and no `value` is set. 
```

* `toArray()`
  * Iterates through remaining cursor value(s) and returns an array of returned row objects. 
* `one()`

>   * Returns a row object if query result has exactly one row. If query result has zero or more than one row, `one()` returns an error. 

Reviewer: ```suggestion
  * Returns a row object if query result has exactly one row. If query result has zero rows or more than one row, `one()` throws an exception. 
```

* `raw()`: Iterator
  * Returns an Iterator over the same query results, with each row as an array of column values (with no column names) rather than an object. 
  * Returned Iterator supports `next()`, `toArray()`, and `one()` methods above.
  * Returned cursor and `raw()` iterator iterate over the same query results and can be combined.

```ts
  let cursor = this.sql.exec("SELECT * FROM artist ORDER BY artistname ASC;");

>   let result = cursor.raw().next();

Reviewer: You may want to call this rawResult.

  if (!result.done) {

>     console.log(result.value); // prints [ artistid: 123, artistname: 'Alice' ]

Reviewer: ```suggestion
    console.log(result.value); // prints [ 123, 'Alice' ]
```

  } else {
    // query returned zero results
  }

>   console.log(cursor.toArray()); // prints [{ artistid: 456, artistname: 'Bob' },{ artistid: 789, artistname: 'Charlie' }]

Reviewer: I get that this example is showing multiple things, but it seems kinda weird to have cursor.toArray() in this example.  It would be better to have individual examples.

Author: > Returned cursor and `raw()` iterator iterate over the same query results and can be combined.

Wanted this example embedded here to highlight the above bullet.

There are individual examples for different tasks below on this page

```

> * `columnNames()`: string[]

Reviewer: columnNames is a property, not a method.

```suggestion
* `columnNames`: string[]
```

Author: I see the below in workers-types. Is this changing with the updated APIs?

```
declare abstract class SqlStorageCursor {
  ...
  get columnNames(): string[];
  ...
}```

>   * Consumes the returned cursor and returns an array of column name strings. Often used with `raw()` to find index of specific column(s).

Reviewer: ```suggestion
  * The column names of the query in the order they appear in each row array returned by the `raw` iterator.
```

> *  `rowsRead()`: number

Reviewer: `rowsRead` is a property, not a method.

```suggestion
*  `rowsRead`: number
```

>   * Returns the number of rows read so far as part of this SQL `query`. This may increase as you iterate the cursor. The final value is used for [SQL billing](/durable-objects/platform/pricing/#sql-storage-billing).

Reviewer: ```suggestion
  * The number of rows read so far as part of this SQL `query`. This may increase as you iterate the cursor. The final value is used for [SQL billing](/durable-objects/platform/pricing/#sql-storage-billing).
```

> *  `rowsWritten()`: number

Reviewer: `rowsWritten is a property, not a method.

```suggestion
*  `rowsWritten`: number
```

>   * Returns the number of rows written so far as part of this SQL `query`. This may increase as you iterate the cursor. The final value is used for [SQL billing](/durable-objects/platform/pricing/#sql-storage-billing).

Reviewer: ```suggestion
  * The number of rows written so far as part of this SQL `query`. This may increase as you iterate the cursor. The final value is used for [SQL billing](/durable-objects/platform/pricing/#sql-storage-billing).
```

#### Examples

<Render file="durable-objects-sql" />

### sql.databaseSize

<code>ctx.storage.sql.databaseSize</code> : number

#### Returns
The current SQLite database size in bytes.

```ts
  let size = ctx.storage.sql.databaseSize;
```

### Point in time recovery

> For [Durable Objects classes with SQL storage](/durable-objects/reference/durable-objects-migrations/#enable-sql-storage-on-new-durable-object-class-migration), the following point-in-time-recovery (PITR) API methods are available to restore a Durable Object's embedded SQLite database to any point in time in the past 30 days. These methods apply to the entire SQLite database contents, including both the object's stored SQL data and stored key-value data using the key-value `put()` API. PITR API is not supported in local development because a durable log of data changes is not stored locally. 

Reviewer: ```suggestion
For [Durable Objects classes with SQL storage](/durable-objects/reference/durable-objects-migrations/#enable-sql-storage-on-new-durable-object-class-migration), the following point-in-time-recovery (PITR) API methods are available to restore a Durable Object's embedded SQLite database to any point in time in the past 30 days. These methods apply to the entire SQLite database contents, including both the object's stored SQL data and stored key-value data using the key-value `put()` API. The PITR API is not supported in local development because a durable log of data changes is not stored locally. 
```

> The point-in-time-recovery (PITR) API represents points in times using "bookmarks". A bookmark is a mostly alphanumeric string like `0000007b-0000b26e-00001538-0c3e87bb37b3db5cc52eedb93cd3b96b`. Bookmarks are designed to be lexically comparable: a bookmark representing an earlier point in time compares less than one representing a later point, using regular string comparison.

Reviewer: ```suggestion
The PITR API represents points in times using "bookmarks". A bookmark is a mostly alphanumeric string like `0000007b-0000b26e-00001538-0c3e87bb37b3db5cc52eedb93cd3b96b`. Bookmarks are designed to be lexically comparable: a bookmark representing an earlier point in time compares less than one representing a later point, using regular string comparison.
```

<code>ctx.storage.getCurrentBookmark()</code> : Promise\<string>

  * Returns a bookmark representing the current point in time in the object's history.

<code>ctx.storage.getBookmarkForTime(timestamp: number | Date)</code>: Promise\<string>

  * Returns a bookmark representing approximately the given point in time, which must be within the last 30 days. If the timestamp is represented as a number, it is converted to a date as if using `new Date(timestamp)`.

<code>ctx.storage.onNextSessionRestoreBookmark(bookmark: string)</code>: Promise\<string>

>   * Instructs that the next time the Durable Object restarts, it should restore its storage to exactly match what it contained at the given bookmark. After calling this, the application should typically invoke `ctx.abort()` to restart the Durable Object, thus completing the point-in-time recovery.

Reviewer: ```suggestion
  * Configure the Durable Object so that the next time it restarts, it should restore its storage to exactly match what the storage contained at the given bookmark. After calling this, the application should typically invoke `ctx.abort()` to restart the Durable Object, thus completing the point-in-time recovery.
```
